configuration/config2/EnvironmentNode.hpp

```diff
@@ -0,0 +1,81 @@
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef _CONFIG2_ENVIRONMENTNODE_HPP_
+#define _CONFIG2_ENVIRONMENTNODE_HPP_
+
+#include <memory>
+#include <map>
+#include "SchemaItem.hpp"
+#include "EnvironmentValue.hpp"
+#include "SchemaValue.hpp"
+#include "Status.hpp"
+#include "ValueDef.hpp"
+#include "platform.h"
+
+
+class DECL_EXPORT EnvironmentNode : public std::enable_shared_from_this<EnvironmentNode>
+{
+    public:
+
+        EnvironmentNode(const std::shared_ptr<SchemaItem> &pCfgItem, const std::string &elemName, const std::shared_ptr<EnvironmentNode> &pParent = nullptr) :
+            m_pSchemaItem(pCfgItem), m_name(elemName), m_pParent(pParent) { }
+        ~EnvironmentNode() { }
+        const std::string &getName() const { return m_name;  }
+        void addChild(std::shared_ptr<EnvironmentNode> pNode);
+        bool removeChild(std::shared_ptr<EnvironmentNode> pNode);
+        std::vector<std::shared_ptr<EnvironmentNode>> getChildren(const std::string &name="") const;
+        std::map<std::string, std::vector<std::shared_ptr<EnvironmentNode>>> getChildrenByName() const;
+        std::map<std::string, std::vector<std::shared_ptr<EnvironmentNode>>> getChildrenByConfigType() const;
+        bool hasChildren() const { return m_children.size() != 0; }
+        int getNumChildren() const { return m_children.size(); }
+        std::shared_ptr<EnvironmentNode> getParent() const;
+        void addAttribute(const std::string &name, std::shared_ptr<EnvironmentValue> pValue);
+        void setAttributeValues(const std::vector<ValueDef> &values, Status &status, bool allowInvalid, bool forceCreate);
+        void setAttributeValue(const std::string &name, const std::string &value, Status &status, bool allowInvalid=false, bool forceCreate=false);   // candidate for a variant?
+        std::string getAttributeValue(const std::string &name) const;                                  // candidate for a variant?
+        void addMissingAttributesFromConfig();
+        bool setNodeValue(const std::string &value, Status &status, bool force = false);
+        void setNodeEnvValue(const std::shared_ptr<EnvironmentValue> &pEnvValue) { m_pNodeValue = pEnvValue;  }
+        const std::shared_ptr<EnvironmentValue> &getNodeEnvValue() const { return m_pNodeValue;  }
+        bool isNodeValueSet() const { return m_pNodeValue != nullptr; }
+        std::vector<std::shared_ptr<EnvironmentValue>> getAttributes() const;
```

Comment:
It's interesting, at least in your current usage you go through the expense of creating a vector only to use the iterator, when you could have iterated over the map.  Maybe you should just support iteration directly, rather than exposing a vector or a map?

Reply:
The reason for a vector is to simplify use of the attributes. It's maintained as a map to aid in enforcing unique attribute names (checked during insertion). At one point I had defined a method that filtered the attributes, thus reducing what is returned to a set smaller than the actual list of attributes.

Reply:
If this is going to be used a lot then I think the name at least is misleading.  I wouldn't think I was doing something as expensive as iterating over the map to create a new vector object.  With my classes at least if a method says "getXXXX" it is usually a very light weight "accessor" function.

So if you want to create a vector becuase 
1. it will be used more extensively than just iteration, 
2. you don't want to have to synchronize around the internal map iteration.
that's fine, but I would give it a more expensive sounding name.

On the other hand if you think it would be safe to provide an iterator instead, that would be more efficient.